### ISSUE TITLE
.on('post-header.bs.table' causing multiple refreshes

### ISSUE THREAD

**Issue by wexwell** [wenzhixin/bootstrap-table#1998]
I had previously posted on this issues before I had isolated it more. I tried posting on the top here, but  I presume no responses due to the fact that it is closed. To recap the situation. Every time a drop select2 box is select the code auto refreshes the datatable. Every time a box is clicked the number of requests doubles (see the previous issue for more details).

https://github.com/wenzhixin/bootstrap-table/issues/1871

However, I was able to further isolate the problem to a specific part of code. I moved one of the boxes out the table / table header and it is no longer submitting everything double. Any box inside of the following statement still submits double each time. (I tried changing it so I am not initializing the event with a class and use the id instead. this made no difference).

Statement
$('#main').on('post-header.bs.table', function () {
     //code to initialize select2 boxes goes here
     //code to auto refresh table on any select2 box close goes here
 });

Website
http://tinkeringtester.com/swtor/stats/dps.html

**Comment by dabros** [comment]
@wexwell - my approach would be to debug js panel in realtime (breakpoints, ect) and see which events triggered.

For that purpose, and so changes can be observed more easily in real time and shared, please create a jsfiddle for this that uses non-minified bootstrap-table.js (and any other relevant js resources, just in case)

https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#bug-reports

That link includes example fiddles that can help.

Data doesnt matter, its just somewhere i can more easily debug and if needed fork/update and include inline copy of any changed files.

Might be a little difficult, but worth the hassle as you assumingly can use the same urls that the server-side examples do - just fork and scale up one of those examples to match minimum use case requirements.

Thanks

**Comment by wexwell** [comment]
The reason it is not in a jsfildde or codeply is because in order to duplicate this it needs to be pulling data from a database. This is the sole purpose I put together a mock-up website the duplicates the problem. Those services seem to block that type of activity.

**Comment by dabros** [comment]
@wexwell - as i said, look at example fiddles which include server side pagination - dummy data is fine, but if we cant make our changes stick then we cant prove any **PR** so we need a fiddle or similar.

**Comment by wexwell** [comment]
The examples are only mock pulling data from a database. This is dynamic and each option changes depending on the values of the others. I have not been able to replicate that without actually querying from a database.

**Comment by dabros** [comment]
@wexwell 

https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#bug-reports

> Other templates can also be found at [jsFiddle Bootstrap](http://bootstrap-table.wenzhixin.net.cn/examples/#basic)

then go down till you see 'pagination' and 'server side' (or crtl+f "server side")

**Comment by wexwell** [comment]
The examples are still pulling data from file sources and not an actual database. Now that I know I can simply move the dropdown boxes out of the table header (for a complete fix) it isn't nearly as clean, but it is a much simpler solution (which has already been implemented in about half of my pages, and the speed increase is VERY significant).

I have put a great deal of time into getting a working example and fixing this. In order to get a 'proper working example,' I am going to have to query select2 json data after it is already completely loaded. I will do that, but I am not able to invest in that right now. Give me a little time.

**Comment by dabros** [comment]
@wexwell - whenever your ready, but as ive said repeatedly now, there IS an example using server data, if you cant use same data endpoints you are in your swotor demo. It is called 'pagination'->'server side', not one of the 3 dummy templates but from the link i quoted above.

Read my last post, im not going to post fiddle link itself as these might change so go to list page.

**Comment by wexwell** [comment]
The example using server data actually references and pulls up a data file from the users directory. When I googled and looked up the syntax of that command the jsfillde does all supported it referencing a file from the users directory.

Server side pagination, from my research, is similar. Declaring variables inside jsfiddle to mimic server side... If I am wrong please direct me to a specific example.

**Comment by dabros** [comment]
@wexwell - read 2 posts ago

---

https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#bug-reports

> Other templates can also be found at [jsFiddle Bootstrap](http://bootstrap-table.wenzhixin.net.cn/examples/#basic)

then go down till you see 'pagination' and 'server side' (or crtl+f "server side")

---

http://jsfiddle.net/4r6g4cfu/3/

**Comment by wexwell** [comment]
His table is initialized in the html portion, mine is initialized using javascript. now, for the purpose of the table I could change the syntax for that 1 line. however, that does not help me with the select2 boxes which have to use ajax (or I would have made the simple revamp.

**Comment by dabros** [comment]
@wexwell - when i said mimic or dummy i refered to jsfiddle AJAX REQUESTS options (see left side)

http://doc.jsfiddle.net/use/echo.html#json

It seems to be in mootools, so i converted to jquery for you.

http://jsfiddle.net/dabros/jtg4pc2v/

Use 'echo' to return whatever data you need, however complex or simple.

You can setup bootstrap-table to use custom ajax (http://issues.wenzhixin.net.cn/bootstrap-table/#options/custom-ajax.html) but current demo seems broken atm and in any case simpler to just use that server-side pagination examples url and mimic whatever you need for ajax requests for select2 to make it meet min requirements.

**Comment by dabros** [comment]
@wexwell  - give me a few min to see if i can get echo to work for data-url

**Comment by dabros** [comment]
So i couldnt see why it wouldnt till i dug into `initServer`, but the json.stringify for '/echo/json/' maybe caused issues as it may have been double encoded.

To avoid this i just changed `contentType` on table itself to avoid this extra logic.

It doesnt matter if searches or page change actually change the returned content, just that the data requests are made.

http://jsfiddle.net/dabros/jtg4pc2v/1/

I override `queryOptions` to avoid any issue with extra param and because `ajaxOptions` didnt want to work in quite the same way, again just because of `echo` requirements.

---

Though i still dont know why you dont just stick a CORS (cross domain request) allowance for github on your data source.

One extra header and we avoid all this hassle, and use exact same code in jsfiddle.

**Comment by wexwell** [comment]
Thank you for for doing that.

The thing that is missing, and the reason I didn't go through making the table yet, is that I cannot duplicate the select2 actions without having a similar solution for the select2 code. the select2 in the row header is what is triggering the multiple refreshes and the problems, and only when it is in the following code (putting it outside of this completely solves the problem, but is not nearly as 'clean' of a product).

$('#main').on('post-header.bs.table', function () {
//code to initialize select2 boxes goes here
//code to auto refresh table on any select2 box close goes here
});

In order to implement the select2 and duplicate the problem, I have to figure out how to make the select2 query a pre-loaded json instead of querying a database and getting the same result set. I can simplify it and have a total of 2 select2 boxes, but I need at least 2 (one inside the header and one outside the header).

**Comment by dabros** [comment]
@wexwell - see the second 'echo' example, you can make it echo back any data in any structure/format you want.

```javascript
$('#findclass').select2({
		width: '100%',
		placeholder: 'Class',
		allowClear: true,
		ajax: {
			url: '/swtor/Includes/selectdpsclass.php',
			dataType: 'json',
			delay: 250,
			data: function (params) {
				$('#findclass').empty();
				var d = new Date();
				var n = d.getTime();
				return {
					q: params.term,
					faction: $('#findfaction').text(),
					n: n
				};
			},
			processResults: function (data) {
				return {
					results: data
				};
			}
		}
	});
```

That should work fine with just replacing the ajax url with '/echo/json/' then passing the data you want it to echo back.

As i said, the actual data is irrelevant in a way, doesnt matter if result set doesnt 'actually' change so long as events are fired off.

**Comment by wexwell** [comment]
If I'm not worrying about parsing the data according to the changes and having and 'actual' working example then I will get the together tomorrow morning. This should have a good enough foundation to allow the events to take place if I implement the rest.

**Comment by dabros** [comment]
@wexwell - you still never answered why you cant just add "Allow-Access-Origin/Allow-Access-Control-Origin" header to your data source - that would avoid all this hassle and allow it to be called from jsfiddle with exact same code as your other demo.

**Comment by wexwell** [comment]
It is not something I am very familiar with, and not sure how it would solve all this.

**Comment by dabros** [comment]
@wexwell - set table url to your swotor url, basically just adding 'http://tinkeringtester.com/' at start of all urls.

You will see error about "Allow-Access-Origin" or similar error in network panel.

Google it

I will bet you money that you will spend not enough a 10th of the time to allow jsfiddle run EXACTLY like your demo (with only exception being moving js into jspanel, scripts into external resource links) as opposed to making this mocku-demo.

Ive said that from the start, i just didnt realise you had no comprehension of the concept - i thought when you said "extensive time spent" it included that.

Basically, the single only reason we need '/echo/' at all is if your original data source doesnt allow "Cross Domain Requests", but thats solvable by a single http header (`<?php header(..); ?>`) being sent from your data source.

At this point we already spent time working around this, but it would still be far easier if you just expose your data source via `*` or to explicit jsfiddle domain. It might take some tweaking, but far less than this whole mess will.

**Comment by dabros** [comment]
@wexwell - ill give you a head start with fiddle at least mostly setup so it should work if you enable cors, but dont wait on that to start googling how to add that header to your server code.

**Comment by dabros** [comment]
@wexwell - here it is

http://jsfiddle.net/dabros/h0u7bhpg/

That is literally how easy it is to make fiddle using your exact data, the only roadblock now is the errors in console and network panel regarding origin headers.

Just google those, not hard to fix at all

**Comment by wexwell** [comment]
This should be together enough to duplicate the problem. I did not put it into a debugging mode, but if you click on the Class button multiple times you will see that it becomes 'buggy' in response. That a symptom of the duplicating refreshes, it becomes responsive when it finishes its multiple refreshes (adding the 1-second delay makes it apparent, good thinking).

http://jsfiddle.net/WexWell/7mp5Lf9c/

Adding HTML:// still isn't getting it to pull data (I had tried that previously, I thought there was more to it than that).

**Comment by dabros** [comment]
@wexwell - you still havent added/changed the server response header (on your server code that these urls point at)

hit f12 and see console errors when running fiddle

**Comment by wenzhixin** [comment]
I think this can solve the problem:
```js
$('#main').on('post-header.bs.table', function () {
  $('.select2').each(function () {
    if ($(this).data('isInit',true)) {
      $(this).data('isInit', true);
      $(this).select2({...});
    } 
  });
});
```
